### ISSUE TITLE
input automaton is too large: 1001 in Operations.topoSortStatesRecurse(Operations.java:1357)

### ISSUE THREAD

**Issue by veita** [apache/lucene#12284]
### Description

The error below appears many times per day and quickly fills the Solr logs up to gigabytes in size.

Unfortunately I cannot tell what kind of input or configuration causes the error. However, the issue seems to be similar to https://github.com/apache/lucene/issues/11809.

```
ERROR 2023-05-09T18:31:56,633Z - org.apache.solr.servlet.HttpSolrCall[qtp1441070244-500081]
 java.lang.IllegalArgumentException: input automaton is too large: 1001
java.lang.IllegalArgumentException: input automaton is too large: 1001
	at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1349) ~[lucene-core-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1357) ~[lucene-core-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1357) ~[lucene-core-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1357) ~[lucene-core-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1357) ~[lucene-core-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1357) ~[lucene-core-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	... 990 identical lines
	at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1357) ~[lucene-core-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1357) ~[lucene-core-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1357) ~[lucene-core-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1357) ~[lucene-core-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1357) ~[lucene-core-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	at org.apache.lucene.util.automaton.Operations.topoSortStatesRecurse(Operations.java:1357) ~[lucene-core-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	at org.apache.lucene.util.automaton.Operations.topoSortStates(Operations.java:1325) ~[lucene-core-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	at org.apache.lucene.search.suggest.analyzing.AnalyzingSuggester.replaceSep(AnalyzingSuggester.java:278) ~[lucene-suggest-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	at org.apache.lucene.search.suggest.analyzing.AnalyzingSuggester.toAutomaton(AnalyzingSuggester.java:877) ~[lucene-suggest-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	at org.apache.lucene.search.suggest.analyzing.AnalyzingSuggester.build(AnalyzingSuggester.java:417) ~[lucene-suggest-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	at org.apache.lucene.search.suggest.Lookup.build(Lookup.java:175) ~[lucene-suggest-9.3.0.jar:9.3.0 d25cebcef7a80369f4dfb9285ca7360a810b75dc - ivera - 2022-07-25 12:30:23]
	at org.apache.solr.spelling.suggest.SolrSuggester.build(SolrSuggester.java:175) ~[solr-core-9.1.1.jar:9.1.1 d998e63978abfedde3b75bab4ba6e1e78ddb5944 - magibney - 2023-01-17 19:58:00]
	at org.apache.solr.handler.component.SuggestComponent.prepare(SuggestComponent.java:197) ~[solr-core-9.1.1.jar:9.1.1 d998e63978abfedde3b75bab4ba6e1e78ddb5944 - magibney - 2023-01-17 19:58:00]
	at org.apache.solr.handler.component.SearchHandler.handleRequestBody(SearchHandler.java:384) ~[solr-core-9.1.1.jar:9.1.1 d998e63978abfedde3b75bab4ba6e1e78ddb5944 - magibney - 2023-01-17 19:58:00]
	at org.apache.solr.handler.RequestHandlerBase.handleRequest(RequestHandlerBase.java:224) ~[solr-core-9.1.1.jar:9.1.1 d998e63978abfedde3b75bab4ba6e1e78ddb5944 - magibney - 2023-01-17 19:58:00]
	at org.apache.solr.core.SolrCore.execute(SolrCore.java:2865) ~[solr-core-9.1.1.jar:9.1.1 d998e63978abfedde3b75bab4ba6e1e78ddb5944 - magibney - 2023-01-17 19:58:00]
	at org.apache.solr.servlet.HttpSolrCall.execute(HttpSolrCall.java:887) ~[solr-core-9.1.1.jar:9.1.1 d998e63978abfedde3b75bab4ba6e1e78ddb5944 - magibney - 2023-01-17 19:58:00]
	at org.apache.solr.servlet.HttpSolrCall.call(HttpSolrCall.java:606) ~[solr-core-9.1.1.jar:9.1.1 d998e63978abfedde3b75bab4ba6e1e78ddb5944 - magibney - 2023-01-17 19:58:00]
	at org.apache.solr.servlet.SolrDispatchFilter.dispatch(SolrDispatchFilter.java:250) ~[solr-core-9.1.1.jar:9.1.1 d998e63978abfedde3b75bab4ba6e1e78ddb5944 - magibney - 2023-01-17 19:58:00]
	at org.apache.solr.servlet.SolrDispatchFilter.lambda$doFilter$0(SolrDispatchFilter.java:218) ~[solr-core-9.1.1.jar:9.1.1 d998e63978abfedde3b75bab4ba6e1e78ddb5944 - magibney - 2023-01-17 19:58:00]
	at org.apache.solr.servlet.ServletUtils.traceHttpRequestExecution2(ServletUtils.java:257) ~[solr-core-9.1.1.jar:9.1.1 d998e63978abfedde3b75bab4ba6e1e78ddb5944 - magibney - 2023-01-17 19:58:00]
	at org.apache.solr.servlet.ServletUtils.rateLimitRequest(ServletUtils.java:227) ~[solr-core-9.1.1.jar:9.1.1 d998e63978abfedde3b75bab4ba6e1e78ddb5944 - magibney - 2023-01-17 19:58:00]
	at org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:213) ~[solr-core-9.1.1.jar:9.1.1 d998e63978abfedde3b75bab4ba6e1e78ddb5944 - magibney - 2023-01-17 19:58:00]
	at org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:195) ~[solr-core-9.1.1.jar:9.1.1 d998e63978abfedde3b75bab4ba6e1e78ddb5944 - magibney - 2023-01-17 19:58:00]
	at org.eclipse.jetty.servlet.FilterHolder.doFilter(FilterHolder.java:201) ~[jetty-servlet-9.4.48.v20220622.jar:9.4.48.v20220622]
	at org.eclipse.jetty.servlet.ServletHandler$Chain.doFilter(ServletHandler.java:1626) ~[jetty-servlet-9.4.48.v20220622.jar:9.4.48.v20220622]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:552) ~[jetty-servlet-9.4.48.v20220622.jar:9.4.48.v20220622]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) ~[jetty-server-9.4.48.v20220622.jar:9.4.48.v20220622]
```

### Version and environment details

Lucene version is 9.3.0
Java version is 17.0.4.1+1

**Comment by tang-hi** [comment]
It seems like the automaton has become too large and has exceeded the recursion limit. Perhaps we should consider changing topoSort to a non-recursive approach. I will work on fixing it when I have some time.

**Comment by tang-hi** [comment]
it should be solved in PR #12286 😆

**Comment by mikemccand** [comment]
Can we close this issue now?  The fix will be released in Lucene 9.7.0.
